Fix part of #8423: New linter check to enforce snake_case

sajalasati opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #[8423](https://github.com/oppia/oppia/issues/8423).
2. This PR does the following:
Ensure that **snake_case** naming convention is followed in all the Python files for **variables**. I have modified `.pylintrc` file by using the following regex for snake_case :
```
variable-rgx=^(?:|_)[a-z][a-z0-9]*(_[a-z0-9]+)*$   (FINAL)
```
Explanation of regex:
- First part `^(?:|_)` is the OR condition - handling whether we start with a `_` or not.
- Second part `[a-z]` means next character after the first part must be one of these characters, because variable names must start with an english alphabet.
- Third part `[a-z0-9]*` means next zero or more of these characters can follow next.
- Fourth part allows zero or more strings of pattern type `(_[0-9a-z]+)` to attach next.
- `$` in the end enforces that pattern must end after fourth part.

3. Also modified one variable each in `scripts/docstrings_checker.py` and `scripts/linters/pylint_extensions.py` which did not follow snake_case naming style.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

Commit:
pylint builtin snake_case enforced

Commit:
Merge remote-tracking branch 'upstream/develop' into new-linter-check

vinitamurthi commented:
@kevinlee12 please change the label if that isn't the right one! Considering that its a linter fix, I added it in the dev ex changelog

kevinlee12 commented:
@vinitamurthi the label is correct.

Hudda commented on .pylintrc:
```diff
@@ -17,10 +17,10 @@ docstring-quote=double
 
 # Regular expression which should only match correct function names
-function-rgx=^[_a-z][a-z0-9_]*$
+function-rgx=^[a-z_][a-z0-9]+(_[a-z0-9]+)*$
```
You don't need to change this one, You only need to change regex for variables. Revert it back to its original form.

sajalasati replied:
Done. Reverted back to original.

Hudda commented on .pylintrc:
```diff
@@ -17,10 +17,10 @@ docstring-quote=double
 
 # Regular expression which should only match correct method names
-method-rgx=^([_a-z][a-z0-9_]*|__[a-z0-9]+__)$
+method-rgx=^[a-z_][a-z0-9]+(_[a-z0-9]+)*$
```
Ditto as above

sajalasati replied:
Done. Reverted back to original.

Hudda commented on .pylintrc:
```diff
@@ -32,13 +32,13 @@ good-names=e,_,d,f,i,l,p,w,fn,fs,id,pc,sc,zf,setUp,tearDown,longMessage,maxDiff
 
 # Regular expression which should match good variable names
-variable-rgx=[a-z_][a-z0-9_]*$
+variable-rgx=^[a-z_][a-z0-9]+(_[a-z0-9]+)*$
```
You need to rethink this regex, this is incorrect.

sajalasati replied:
Now it can handle single character variable names too. Variable name must start with an alphabet. If `_` is the first character then the second literal must be an alphabet. I hope the regex looks correct now.

Hudda commented on .pylintrc:
```diff
@@ -32,13 +32,13 @@ good-names=e,_,d,f,i,l,p,w,fn,fs,id,pc,sc,zf,setUp,tearDown,longMessage,maxDiff
 
 # Regular expression matching correct attribute names
-attr-rgx=[A-Za-z_][A-Za-z0-9_]*$
+attr-rgx=^[a-z_][a-z0-9]+(_[a-z0-9]+)*$
```
No need to change this one.

sajalasati replied:
Done. Reverted back to original.

Hudda commented on .pylintrc:
```diff
@@ -32,13 +32,13 @@ good-names=e,_,d,f,i,l,p,w,fn,fs,id,pc,sc,zf,setUp,tearDown,longMessage,maxDiff
 
 # Regular expression matching correct argument names
-argument-rgx=[A-Za-z_][A-Za-z0-9_]*$
+argument-rgx=^[a-z_][a-z0-9]+(_[a-z0-9]+)*$
```
Ditto as above.

sajalasati replied:
Done. Reverted back to original.

Hudda reviewed (approved):
Thanks @sajalasati, I've left some comments PTAL!

You need to change just variable regex.

Commit:
regex changed acc to comments

Commit:
Merge remote-tracking branch 'upstream/develop' into new-linter-check

sajalasati commented:
Are variable names ending with an underscore considered as snake_case? Like the current regex for variable names couldn't match to `property_`, `id_`. @Hudda any comment on this? Should I change the regex to handle a single trailing underscore?

kevinlee12 commented:
@sajalasati technically no.

sajalasati commented:
> @sajalasati technically no.

@kevinlee12 
Then how do you think I can proceed? Create a separate issue to fix those names? or else how?

kevinlee12 commented:
actually, I take my comment back.

@oppia/core-maintainers do you guys know if this is a pattern in our codebase?

seanlip commented:
@sajalasati, where are you seeing those names exactly? Can you point us with links to the relevant files in the codebase? In general, try to provide as much concrete detail/evidence as you can on the first go, otherwise lots of back and forth will be needed.

(I've always thought that we put underscores at the beginning of names to denote private things, so I'm not sure I understand the underscore-at-the-end-of-things issue. Hence the need for evidence and to see the actual usages. Also how many of these are there?)

sajalasati commented:
@seanlip Sorry for that!

The linter errors came at the following places:

- Module: oppia.scripts.docstrings_checker
    - [85, 4](https://github.com/oppia/oppia/blob/develop/scripts/docstrings_checker.py#L85): Variable name "property_" doesn't conform to `u'^(?:([a-z][a-z0-9]*)|(_[a-z][a-z0-9]*))(_[a-z0-9]+)*$'` pattern (invalid-name)
    - [93,16](https://github.com/oppia/oppia/blob/develop/scripts/docstrings_checker.py#L93): Variable name "property_" doesn't conform to `u'^(?:([a-z][a-z0-9]*)|(_[a-z][a-z0-9]*))(_[a-z0-9]+)*$'` pattern (invalid-name)
    - But it didn't give error for the [96th line](https://github.com/oppia/oppia/blob/develop/scripts/docstrings_checker.py#L96) of this same file, maybe because it's just a return statement.

- Module: oppia.core.storage.feedback.gae_models_test
    - [105, 8](https://github.com/oppia/oppia/blob/develop/core/storage/feedback/gae_models_test.py#L105): Variable name "id_" doesn't conform to `u'^(?:([a-z][a-z0-9]*)|(_[a-z][a-z0-9]*))(_[a-z0-9]+)*$'` pattern (invalid-name)

- Module: oppia.scripts.linters.pylint_extensions
    - [449,12](https://github.com/oppia/oppia/blob/develop/scripts/linters/pylint_extensions.py#L449): Variable name "property_" doesn't conform to `u'^(?:([a-z][a-z0-9]*)|(_[a-z][a-z0-9]*))(_[a-z0-9]+)*$'` pattern (invalid-name)

Please share your comments as to how should we handle this - by changing the regex or creating separate issue to fix them or change them myself (although I feel you wouldn't be suggesting that).

Explanation of regex for snake case used by me:
- `^(?:([a-z][a-z0-9]*)|(_[a-z][a-z0-9]*))` is an OR statement meaning the name either begins 
    - with a lowercase alphabet followed by zero or more lowercase alphabets/numeric characters
    - or with an underscore, with next character as lowercase alphabet and followed by zero or more lowercase alphabets/numeric characters.
- The leftover part of regex `(_[a-z0-9]+)*$` means whenever an underscore `_` comes again, it should be followed by one or more lowercase characters/numbers. This constitutes one pattern and zero or more such patterns can occur, like we could expect in a general snake case pattern.

Hope this helps!

seanlip commented:
Thanks @sajalasati for the examples. I think you can change these specific variable names to have the underscore at the beginning instead of the end. Would that work? Please keep the regex restrictive so that it does not allow trailing underscores. Thanks!

Commit:
changed variable names

Commit:
Merge remote-tracking branch 'upstream/develop' into new-linter-check

sajalasati commented:
@seanlip I have made the change you suggested as it looked fine to me. Let's see the results of CircleCI tests now.

seanlip reviewed (approved):
LGTM for codeowner file. Thanks!

bansalnitish reviewed (approved):
LGTM for codeowners file!

Hudda commented:
@sajalasati Can you please check if this check is also valid for dictionary keys as mentioned in this [comment](https://github.com/oppia/oppia/issues/8423#issuecomment-644888223).

sajalasati commented:
@Hudda No, the current change won't ensure snake case for dictionary keys. I just checked it by running the linter on some random file after modifying the dictionary keys and it doesn't detect that. I think I will have to write a pylint extension for that. I should be modifying `pylint_extensions.py` [file](https://github.com/oppia/oppia/blob/develop/scripts/linters/pylint_extensions.py) only right? I looked at it a while ago, but didn't complete it. I will push that too in this PR asap (within next 2 days max). Is that fine?

Hudda commented:
Yep, that's fine.

sajalasati commented:
@Hudda and others
**Rule**: `"follow snake_case style for dictionary keys"`
For my leftover work in this PR of enforcing this above **rule** in all python files, I have encountered two types of cases where we must check/enforce this. I would like to discuss my approach first with some implementation details before pushing the code:
- **Case 1**: when we are creating a dictionary
    - Sub-case: creating and at the same time passing dictionary as an argument.
    ```python
    self.values.update({
                'topic_dict': topic.to_dict(),
                'grouped_skill_summary_dicts': grouped_skill_summary_dicts,
                'skill_id_to_description_dict': skill_id_to_description_dict,
                'skill_id_to_rubrics_dict': skill_id_to_rubrics_dict
        })
    ```
    - Sub-case: assigning a new dictionary to a variable, attribute etc
    ```python
    some_variable = {"key_1": 1, "key_2": 2}
    ```
   
In this case, we will check all the keys of the dictionary for the **rule**. 
**Implementation**: To tackle this case, I am using the function `visit_dict` in the extension to get all the dictionaries - so that I would get all the nodes of Astroid class [Dict](http://pylint.pycqa.org/projects/astroid/en/latest/api/astroid.nodes.html#astroid.nodes.Dict).

__Problem__ - But we do not want to enforce this rule for some files like `pylint_extensions.py`, where a different naming style is followed for certain special dictionary keys. For eg. [here](https://github.com/oppia/oppia/blob/develop/scripts/linters/pylint_extensions.py#L946-L952) and several other places in this file

```
msgs = {
        'C0009': (
            'Importing %s layer in %s layer is prohibited.',
            'invalid-import',
            'Storage layer and domain layer must not import'
            'domain layer and controller layer respectively.'),
    }
```
**So is there a way to skip this lint check for some files like this one? Please tell if there is one, I am unable to find any.**

- **Case 2**: when we are accessing a dictionary key for eg.
```python
    summary['story_is_published'] = (story_id_to_publication_status_map[summary['id']])
``` 
To tackle this case, I am using the function `visit_subscript` inside the extension based on the Astroid node class type [Subscript](http://pylint.pycqa.org/projects/astroid/en/latest/api/astroid.nodes.html#astroid.nodes.Subscript) that I should be needing for this case. 

Although this "Subscript" node class type also gives me instances where the sub scripted part is not a string like `story_id_to_publication_status_map[reference.story_id]`, but I have only enforced my naming style when the key/sub scripted part is an actual string (which are received from the node as a unicode string i.e like `u'xyz string'`).

**My Questions**:
1) The problem mentioned in Case 1 - how to avoid this lint check for some special files?
2) Do the above 2 cases cover all the cases you had in mind?
2) Am I using the correct Node classes to tackle each of the cases?

After my above doubts are resolved, I will first push the code written in `pylint_extensions.py` and then proceed to writing tests for the my new class in the tests file.

Hudda commented:
@sajalasati 
1. Check if you are able to get the name of the variable to which dictionary is assigned for example: `msgs` and ignore that dictionary with variable name `msgs`
2. Yes, these two cover all the cases.
3. These node classes seem fine to me, but I'm not sure about case-2

sajalasati committed:
snake case style check

Added snake case style check for dictionary keys in python

Commit:
solving merge conflict

Hudda commented:
@sajalasati I think you used `rebase` instead of `merge`. Please close this one and open a new PR.

Commit:
Merge remote-tracking branch 'upstream/develop' into new-linter-check

Commit:
snake case for dict keys

Commit:
fix file to disable error

Commit:
fix file to disable error

sajalasati commented:
@Hudda 
I think it's fine now. I merged it with the latest develop and pushed it. I have also pushed the code for the lint check.
To answer your questions:
- Part 1) To resolve this, I have added a disable statement next to all the `msgs` dictionaries in the `pylint_extensions.py` file. 
- Part 3) My explanation for case 2 is that using keys of a dictionary is a sub-part of the sub-scripting thing which astroid [Subscript](http://pylint.pycqa.org/projects/astroid/en/latest/api/astroid.nodes.html#astroid.nodes.Subscript) class covers. Consider the case where the subscript is `name_1[name_2[name_3["key_string"]]]` - what my function `visit_subscript` receives for this is all the subscripts i.e `'name_2[name_3["key_string"]]'`, `'name_3["key_string"]'`, `'u"key_string"'`, and I have checked only for strings beginning with a unicode character. Hence I think it will cover all the cases I can think of. Do you have any other case which the above logic might not cover?

Also in the file `core/controllers/topic_editor.py` and its test file, I had to add a disable lint check statement too, because I think I have to change it at more places than just these 2 files (like maybe in front end related files but I am not sure).

Meanwhile I am working on writing tests for this lint check in the pylint extensions test file.

sajalasati commented:
@Hudda @kevinlee12 

- The lint error "snake case naming convention not followed" is coming on **102 files** - needs **1.4k+** modifications in total across all files. So I am not sure how to move ahead. I created a gist, so you can see some example errors [here](https://gist.github.com/sajalasati/87a88dcc0df6f3fc91252a1eda66b977).
- I think originally snake_case for dictionary keys was meant to be enforced only in backend files (going by [this](https://github.com/oppia/oppia/pull/8868#issuecomment-602290760) and [this](https://github.com/oppia/oppia/pull/8885#issuecomment-602058511) comment added on the original issue page). But I am currently not able to have control on that, even the python files testing the front end part are being affected, which I don't think is the aim.

Should I cover snake case for dictionary keys in this PR or not?

srijanreddy98 commented:
I don't think my review is required here!

Hudda commented:
> @Hudda @kevinlee12
> 
>     * The lint error "snake case naming convention not followed" is coming on **102 files** - needs **1.4k+** modifications in total across all files. So I am not sure how to move ahead. I created a gist, so you can see some example errors [here](https://gist.github.com/sajalasati/87a88dcc0df6f3fc91252a1eda66b977).
> 
>     * I think originally snake_case for dictionary keys was meant to be enforced only in backend files (going by [this](https://github.com/oppia/oppia/pull/8868#issuecomment-602290760) and [this](https://github.com/oppia/oppia/pull/8885#issuecomment-602058511) comment added on the original issue page). But I am currently not able to have control on that, even the python files testing the front end part are being affected, which I don't think is the aim.
> 
> 
> Should I cover snake case for dictionary keys in this PR or not?

I don't think if it is a good idea to have this check for dictionary keys because we need CamelCase keys at several places like in `pylint_extensions`. @kevinlee12 what do you think?

kevinlee12 commented:
Yup, I agree.

> On Jul 7, 2020, at 8:49 AM, Anshul Hudda <notifications@github.com> wrote:
> 
> ﻿
> @Hudda @kevinlee12
> 
> * The lint error "snake case naming convention not followed" is coming on **102 files** - needs **1.4k+** modifications in total across all files. So I am not sure how to move ahead. I created a gist, so you can see some example errors [here](https://gist.github.com/sajalasati/87a88dcc0df6f3fc91252a1eda66b977).
> 
> * I think originally snake_case for dictionary keys was meant to be enforced only in backend files (going by [this](https://github.com/oppia/oppia/pull/8868#issuecomment-602290760) and [this](https://github.com/oppia/oppia/pull/8885#issuecomment-602058511) comment added on the original issue page). But I am currently not able to have control on that, even the python files testing the front end part are being affected, which I don't think is the aim.
> Should I cover snake case for dictionary keys in this PR or not?
> 
> I don't think if it is a good idea to have this check for dictionary keys because we need CamelCase keys at several places like in pylint_extensions. @kevinlee12 what do you think?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

Hudda commented:
@sajalasati as per discussion we don't need this check for dictionary keys, you can remove it.

Commit:
removed snake case check

Commit:
Merge remote-tracking branch 'upstream/develop' into new-linter-check

Commit:
Merge remote-tracking branch 'upstream/develop' into new-linter-check

sajalasati commented:
@Hudda Reverted back the changes made for dictionary keys. Modified the PR description accordingly. Kindly check now.

Hudda reviewed (approved):
Thanks, @sajalasati, LGTM :)

Hudda commented:
@DubeySandeep PTAL!

Hudda commented:
@sajalasati Please merge from the develop.

Commit:
resolve merge conflict

sajalasati commented:
@Hudda Done!

DubeySandeep commented on core/storage/feedback/gae_models_test.py:
```diff
@@ -94,6 +94,87 @@ def test_has_reference_to_user_id(self):
             feedback_models.GeneralFeedbackThreadModel
             .has_reference_to_user_id(self.NONEXISTENT_USER_ID))
 
+    def test_get_user_id_migration_policy(self):
```
Why these changes? Doesn't seems like part of this PR as per PR title.

sajalasati replied:
I had made some modifications to this function earlier, which were needed to ensure snake case naming style, but the function was deleted. So it came alongwith my other changes and I didn't notice it. Removed it now.

DubeySandeep commented on scripts/docstrings_checker.py:
```diff
@@ -82,18 +82,18 @@ def get_setters_property(node):
         the given setter node, or None if one could not be found.
     """
-    property_ = None
+    _property = None
```
This is not a private value, right? `property` is a keyword and that's why we have used _ at the end. I would suggest changing this to `setters_property`, will that be fine?

sajalasati replied:
Done!

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -568,9 +568,9 @@ def visit_raise(self, node):
-            property_ = docstrings_checker.get_setters_property(func_node)
-            if property_:
-                func_node = property_
+            _property = docstrings_checker.get_setters_property(func_node)
```
Ditto*

sajalasati replied:
Done!

DubeySandeep commented on .pylintrc:
```diff
@@ -32,7 +32,7 @@ good-names=e,_,d,f,i,l,p,w,fn,fs,id,pc,sc,zf,setUp,tearDown,longMessage,maxDiff
 
 # Regular expression which should match good variable names
-variable-rgx=[a-z_][a-z0-9_]*$
+variable-rgx=^(?:([a-z][a-z0-9]*)|(_[a-z][a-z0-9]*))(_[a-z0-9]+)*$
```
This looks a bit complicated. FYI, pylint allows using `<snake>`  in the regex ([ref](http://pylint.pycqa.org/en/latest/user_guide/options.html?highlight=snake#cmdoption-inlinevar-rgx)), can we make it simpler using that? (not sure!)

sajalasati replied:
Done.

sajalasati replied:
Also the builtin-snake case naming convention allows us to have variable names with trailing underscores, like `property_` which I have changed in 2 files - should I revert back those names or are those changes fine?

seanlip replied:
Please forbid trailing underscores -- we don't use these, as a rule. Thanks!

sajalasati replied:
I have used built-in snake case convention, and it does allow to have trailing underscore. You mean I should make changes to forbid that? Also this built-in one (the way I have used it) disallows single/double length variable names - gave the following linter errors:
```
************* Module oppia.scripts.release_scripts.wrap_up_release
C: 68, 8: Variable name "pr" doesn't conform to snake_case naming style 
C: 72, 8: Variable name "pr" doesn't conform to snake_case naming style 
C:145, 4: Variable name "g" doesn't conform to snake_case naming style
```

My own regex `^(?:([a-z][a-z0-9]*)|(_[a-z][a-z0-9]*))(_[a-z0-9]+)*$` forbids trailing underscores as well as allows above names. It passed all the linter tests also earlier.
@DubeySandeep 
How should I proceed? Use my own regex or try some modification with built-in snake case only?

**Edit** - Logic for my regex:
- First part `^(?:([a-z][a-z0-9]*)|(_[a-z][a-z0-9]*))` is the OR condition - handling whether we start with a `_` or not.
- Second part allows zero or more strings of pattern type `(_[0-9a-z]+)` to attach to the first part.

DubeySandeep replied:
Hi @sajalasati, Thanks for checking this, I'm fine with having the regex you have written but just too simplify it can you check whether we can use `<snake>` in the regex to make it simpler? (You can check [this](http://pylint.pycqa.org/en/latest/user_guide/options.html?highlight=snake#cmdoption-inlinevar-rgx) to see how to use <snake> in the regex)

DubeySandeep replied:
I read [this comment](https://github.com/oppia/oppia/pull/9659#issuecomment-657114639) and it gives answer to my question! Thanks for all the work, @sajalasati!

sajalasati replied:
Ok, that's good! Thanks!

DubeySandeep reviewed (changes requested):
@sajalasati, Thanks for the PR, I've left a few comments PTAL!

 @Hudda, wanted to confirm that this is just a part of `Enforce that PEP8 naming convention is followed. Please check the comment here to understand the use case.` check, right? we still have to ensure that we follow correct patterns for other things, right? (I'm fine with having this PR just focus on the variable name style!)

Hudda commented:
> @Hudda, wanted to confirm that this is just a part of Enforce that PEP8 naming convention is followed. Please check the comment here to understand the use case. check, right?

@DubeySandeep  I think we have just this one use case, but I'm not sure. Can you please provide a few examples?

Commit:
Merge remote-tracking branch 'upstream/develop' into new-linter-check

Commit:
removing removed function in develop

Commit:
fixed name changes

Commit:
fixed lint error

Commit:
built-in function for snake case

Hudda commented:
@DubeySandeep ping!

Hudda commented:
@sajalasati lint tests are failing PTAL!

Commit:
Merge remote-tracking branch 'upstream/develop' into new-linter-check

sajalasati commented:
> @sajalasati lint tests are failing PTAL!

@Hudda Actually I mentioned about the failing tests in [this comment](https://github.com/oppia/oppia/pull/9659#discussion_r452042198). But I will re-write it here.
Using `variable-naming-style=snake_case` in `.pylintrc` file ([ref](http://pylint.pycqa.org/en/latest/user_guide/options.html?highlight=snake#cmdoption-inlinevar-rgx)), I am getting the following lint errors:
```
************* Module oppia.scripts.release_scripts.wrap_up_release
C: 68, 8: Variable name "pr" doesn't conform to snake_case naming style 
C: 72, 8: Variable name "pr" doesn't conform to snake_case naming style 
C:145, 4: Variable name "g" doesn't conform to snake_case naming style
```
I checked it locally and found that it allows trailing underscores, disallows single/double length names as above.

The regex for built-in snake case isn't available, so I am not sure how I can append something to it or modify it.
I used the example in this same [link](http://pylint.pycqa.org/en/latest/user_guide/options.html?highlight=snake#cmdoption-inlinevar-rgx) as @DubeySandeep suggested me to use, i.e using `variable-rgx=^(?P<snake>[a-z_]+)$` but that too allows trailing underscores. I tried doing some modification like `variable-rgx=^(?P<snake>[a-z]+)$` and `variable-rgx=^(?P<snake>[a-z_]*[a-z])$` but they don't work. I am unable to understand the working of built-in snake case.

That's why I want to go ahead with using my own regex: `^(?:|_)[a-z][a-z0-9]*(_[a-z0-9]+)*$`. **I have simplified it compared to what I originally thought**. It forbids trailing underscores as well as allows above names which are shown as linter errors currently. It passed all the linter tests also earlier.
**Logic** for my regex:
- First part `^(?:|_)` is the OR condition - handling whether we start with a `_` or not.
- Second part `[a-z]` means next character after the first part must be one of these characters, because variable names must start with an english alphabet.
- Third part `[a-z0-9]*` means next zero or more of these characters can follow next.
- Fourth part allows zero or more strings of pattern type `(_[0-9a-z]+)` to attach next.
- `$` in the end enforces that pattern must end after fourth part.

@Hudda What are your comments on this?
@DubeySandeep I have simplified my regex now, and pushed changes to actually verify also that it's correct. Will it be okay?

Commit:
snake case regex changed

DubeySandeep commented:
> @DubeySandeep I think we have just this one use case, but I'm not sure. Can you please provide a few examples?

@Hudda, I think that's the part of the task to check where we need to do the changes!

DubeySandeep reviewed (approved):
LGTM!

sajalasati commented:
@Hudda I believe you can merge the PR now if all looks good to you. I have modified the [Overview Section](https://github.com/oppia/oppia/pull/9659#issue-439883880) of PR accordingly so you can check that too.
